File: external/java-impl/src/main/java/armyant/hgoe/inmemory/nodes/EntityNode.java
package armyant.hgoe.inmemory.nodes;

import armyant.hgoe.inmemory.Rankable;
import armyant.structures.Document;

/**
 * Created by jldevezas on 2017-10-24.
 */
public class EntityNode extends Node implements Rankable {
    private String docID;

    public EntityNode() {
    }

    public EntityNode(String name) {
        this(null, name);
    }

    public EntityNode(Document document, String name) {
        super(name);
        if (name.equals(document.getTitle())) {
            this.docID = document.getDocID();
        } else {
            this.docID = null;
        }
    }

    @Override
    public void setDocID(String docID) {
        this.docID = docID;
    }

    @Override
    public String getDocID() {
        return docID;
    }

    public boolean hasDocID() {
        return docID != null;
    }
}

File: external/cpp-impl/src/hgoe/nodes/entity_node.cpp
//
// Created by jldevezas on 3/8/18.
//

#include <hgoe/nodes/entity_node.h>

BOOST_CLASS_EXPORT_IMPLEMENT(EntityNode)

EntityNode::EntityNode() = default;

EntityNode::EntityNode(std::string name) : Node(boost::move(name)) { }

EntityNode::EntityNode(Document *document, std::string name) : Node(name) {
    if (name == document->getTitle()) {
        this->docID = document->getDocID();
    }
}

Node::Label EntityNode::label() const {
    return Label::ENTITY;
}

void EntityNode::setDocID(const std::string &docID) {
    this->docID = docID;
}

std::string EntityNode::getDocID() const {
    return docID;
}

bool EntityNode::hasDocID() {
    return !docID.empty();
}

void EntityNode::print(std::ostream &os) const {
    os << "EntityNode { docID: " << (docID.empty() ?  "N/A" : docID) << " } & ";
    Node::print(os);
}
